hazelcast/src/main/java/com/hazelcast/transaction/impl/xa/XAResourceImpl.java

```diff
@@ -0,0 +1,303 @@
+/*
+ * Copyright (c) 2008-2015, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.transaction.impl.xa;
+
+import com.hazelcast.cluster.ClusterService;
+import com.hazelcast.config.GroupConfig;
+import com.hazelcast.instance.MemberImpl;
+import com.hazelcast.nio.Address;
+import com.hazelcast.nio.serialization.Data;
+import com.hazelcast.partition.InternalPartitionService;
+import com.hazelcast.spi.AbstractDistributedObject;
+import com.hazelcast.spi.InternalCompletableFuture;
+import com.hazelcast.spi.NodeEngine;
+import com.hazelcast.spi.OperationService;
+import com.hazelcast.spi.impl.SerializableCollection;
+import com.hazelcast.transaction.HazelcastXAResource;
+import com.hazelcast.transaction.TransactionContext;
+import com.hazelcast.transaction.TransactionOptions;
+import com.hazelcast.transaction.impl.Transaction;
+import com.hazelcast.util.ExceptionUtil;
+
+import javax.transaction.xa.XAException;
+import javax.transaction.xa.XAResource;
+import javax.transaction.xa.Xid;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.TimeUnit;
+
+import static com.hazelcast.transaction.impl.xa.XAService.SERVICE_NAME;
+
+/**
+ * Server side XaResource implementation
+ */
+public final class XAResourceImpl extends AbstractDistributedObject<XAService> implements HazelcastXAResource {
+
+    private static final int DEFAULT_TIMEOUT = (int) TimeUnit.MILLISECONDS.toSeconds(TransactionOptions.DEFAULT_TIMEOUT_MILLIS);
+
+    private final ConcurrentMap<Long, TransactionContext> threadContextMap = new ConcurrentHashMap<Long, TransactionContext>();
+    private final ConcurrentMap<Xid, List<TransactionContext>> xidContextMap
+            = new ConcurrentHashMap<Xid, List<TransactionContext>>();
+    private final String groupName;
+
+    public XAResourceImpl(NodeEngine nodeEngine, XAService service) {
+        super(nodeEngine, service);
+        GroupConfig groupConfig = nodeEngine.getConfig().getGroupConfig();
+        groupName = groupConfig.getName();
+    }
+
+    @Override
+    public synchronized void start(Xid xid, int flags) throws XAException {
```

Comment:
same here; do all methods have to be `synchronized`? At least you should use an internal mutex, because instance of this class is exposed to the user, one can use it as his/her own monitor object... which can lead to an unexpected deadlock.